src/number_theory/bernoulli_polynomials.lean

```diff
@@ -0,0 +1,175 @@
+## TODO
+
+- `bernoulli_poly_eval_one_neg` : $$ B_n(1 - x) = (-1)^n*B_n(x) $$
+- ``bernoulli_poly_eval_one` : Follows as a consequence of `bernoulli_poly_eval_one_neg`.
+
+-/
+
+noncomputable theory
+open_locale big_operators
+open_locale nat
+
+open nat finset
+
+/-- The Bernoulli polynomials are defined in terms of the negative Bernoulli numbers. -/
+def bernoulli_poly (n : ℕ) : polynomial ℚ :=
+  ∑ i in range (n + 1), polynomial.monomial (n - i) ((bernoulli i) * (choose n i))
+
+lemma bernoulli_poly_def (n : ℕ) : bernoulli_poly n =
+  ∑ i in range (n + 1), polynomial.monomial i ((bernoulli (n - i)) * (choose n i)) :=
+begin
+  rw [←sum_range_reflect, add_succ_sub_one, add_zero, bernoulli_poly],
+  apply sum_congr rfl,
+  rintros x hx,
+  rw mem_range_succ_iff at hx, rw [choose_symm hx, nat.sub_sub_self hx],
+end
+
+namespace bernoulli_poly
+
+/-
+### examples
+-/
+
+section examples
+
+@[simp] lemma bernoulli_poly_zero : bernoulli_poly 0 = 1 :=
+by simp [bernoulli_poly]
+
+@[simp] lemma bernoulli_poly_eval_zero (n : ℕ) : (bernoulli_poly n).eval 0 = bernoulli n :=
+begin
+ rw [bernoulli_poly, polynomial.eval_finset_sum, sum_range_succ],
+  have : ∑ (x : ℕ) in range n, bernoulli x * (n.choose x) * 0 ^ (n - x) = 0,
+  { apply sum_eq_zero (λ x hx, _),
+    have h : 0 < n - x := nat.sub_pos_of_lt (mem_range.1 hx),
+    simp [h] },
+  simp [this],
+end
+
+end examples
+
+@[simp] theorem sum_bernoulli_poly (n : ℕ) :
+  ∑ k in range (n + 1), ((n + 1).choose k : ℚ) • bernoulli_poly k =
+    polynomial.monomial n (n + 1 : ℚ) :=
+begin
+ simp_rw [bernoulli_poly_def, finset.smul_sum, finset.range_eq_Ico, ←finset.sum_Ico_Ico_comm,
+    finset.sum_Ico_eq_sum_range],
+  simp only [cast_succ, nat.add_sub_cancel_left, nat.sub_zero, zero_add, linear_map.map_add],
+  simp_rw [polynomial.smul_monomial, mul_comm (bernoulli _) _, smul_eq_mul, ←mul_assoc],
+  conv_lhs { apply_congr, skip, conv
+    { apply_congr, skip,
+      rw [nat.prime.choose_mul ((nat.le_sub_left_iff_add_le (mem_range_le H)).1 (mem_range_le H_1))
+        (le.intro rfl), add_comm x x_1, nat.add_sub_cancel, mul_assoc, mul_comm, ←smul_eq_mul,
+        ←polynomial.smul_monomial], },
+    rw [←sum_smul], },
+  rw sum_range_succ,
+  simp only [add_right_eq_self, cast_succ, mul_one, cast_one, cast_add, nat.add_sub_cancel_left,
+    choose_succ_self_right, one_smul, bernoulli_zero, sum_singleton, zero_add,
+    linear_map.map_add, range_one],
+  have f : ∀ x ∈ range n, 2 ≤ n + 1 - x,
+  { rintros x H, rw [mem_range] at H,
+    exact nat.le_sub_left_of_add_le (succ_le_succ H) },
+  apply sum_eq_zero (λ x hx, _),
+  rw [sum_bernoulli _ (f x hx), zero_smul],
+end
+
+open power_series
+open polynomial (aeval)
+variables {A : Type*} [integral_domain A] [algebra ℚ A] [char_zero A]
+
+lemma mem_range_ite_sub_zero {n x : ℕ} (hx : x ∈ range n) : ite (n - x = 0) 1 0 = (0 : A) :=
+begin
+  split_ifs,
+   { exfalso, apply mem_range_sub_ne_zero hx h, },
+   { refl, },
+end
+
+theorem exp_bernoulli_poly (t : A) :
```

Comment:
Can this proof reuse results from the PR on Bernoulli power series?
I'm still a bit uneasy about how long this proof is. There should be useful and reusable building blocks that can be used to make this shorter.

Reply:
Just looking at this now. It seems to me that the proof of `bernoulli_power_series_mul_exp_sub_one` in #6456 follows easily from `bernoulli'_power_series_mul_exp_sub_one` -- we don't need to prove both -- but that both of these proofs follow from the bernoulli polynomial result here, which is strictly stronger (`bernoulli_power_series_mul_exp_sub_one` and `bernoulli'_power_series_mul_exp_sub_one` both follow from this result by setting t=0 and t=1).